container/sagemaker/python_service.py

```diff
@@ -0,0 +1,151 @@
+
+    def on_post(self, req, res):
+        context, data = self._parse_http_request(req)
+
+        if self.handler:
+            log.warning('Using handler, input_handler and output_handler will be ignored.')
+            log.info('handler type: ' + str(type(self.handler)))
+            res.status = falcon.HTTP_200
+            res.body = self.handler(data, context)
+        elif self.input_handler and self.output_handler:
+            processed_input = self.input_handler(data, context)
+            response = requests.post(context.rest_uri, data=processed_input)
+            res.status = falcon.HTTP_200
+            res.body, res.content_type = self.output_handler(response, context)
+        else:
+            res.status = falcon.HTTP_500
+            res.body = 'Handlers are not implemented correctly.'
+
+    def _import_handlers(self):
+        # import handlers
+        import importlib.util
+        spec = importlib.util.spec_from_file_location('inference',
+                                                      '/opt/ml/model/inference_example.py')
+        inference = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(inference)
+
+        self.handler = inference.handler
+        self.input_handler = inference.input_handler
+        self.output_handler = inference.output_handler
+
+    def _parse_http_request(self, req):
+        content_type = req.get_header('Content-Type') or DEFAULT_CONTENT_TYPE
+        accept = req.get_header('Accept') or DEFAULT_ACCEPT_HEADER
+        custom_attributes = req.get_header('X-Amzn-SageMaker-Custom-Attributes')
+        model_name = req.get_header('model_name') or TFS_DEFAULT_MODEL
+        model_version = req.get_header('model_version')
+        method = req.get_header('method') or TFS_DEFAULT_METHOD
+
+        uri = self._tfs_rest_uri(self._parse_custom_attributes(custom_attributes))
+        rest_uri = req.get_header('rest_uri') or uri
+        grpc_uri = req.get_header('grpc_uri')
+
+        custom_attributes = self._make_custom_attributes(custom_attributes,
+                                                         model_name,
+                                                         model_version,
+                                                         method)
+        context = Context(model_name=model_name, model_version=model_version, method=method,
+                          rest_uri=rest_uri, grpc_uri=grpc_uri, custom_attributes=custom_attributes,
+                          request_content_type=content_type, accept_header=accept)
+        data = req.stream.read()
+
+        return context, data
+
+    def _make_custom_attributes(self, custom_attributes, model_name, model_version, tfs_method):
+        attribute_list = []
+        if custom_attributes:
+            attribute_list = re.findall("(tfs-[a-z\-]+=[^,]+)",  # pylint: disable=W1401; # noqa: W605, E501
+                                        custom_attributes)
+        name = False
```

Comment:
this code could be simplified a lot. consider named group matches or at least capturing group matches. declare the vars with initial value None instead, and assign a real string value to them if they are found instead of just True.

Reply:
I will modify this.